weave/monitoring/openai/openai.py

```diff
@@ -2,25 +2,29 @@
 
 import functools
 from contextlib import contextmanager
-from typing import Callable, Union
+from functools import partialmethod
+from typing import Callable, Type, Union
 
 import openai
 from openai import AsyncStream, Stream
 from openai.types.chat import ChatCompletion
 from packaging import version
 
-from weave import graph_client_context
-from weave import run_context
+from weave import graph_client_context, run_context
 
+from ..monitor import _get_global_monitor
 from .models import *
 from .util import *
 
-from ..monitor import _get_global_monitor
-
 old_create = openai.resources.chat.completions.Completions.create
 old_async_create = openai.resources.chat.completions.AsyncCompletions.create
 
 
+class partialmethod_with_self(partialmethod):
+    def __get__(self, obj: Any, cls: Optional[Type[Any]] = None) -> Callable:
```

Comment:
Are tracebacks through patched openai call readable now? Should we attach a __name__ to the function we're generating here?

Reply:
I think they are readable -- this is what I would expect for the example code:
```py
import weave
from openai import OpenAI

@weave.op()
def correct_grammar(sentence: str) -> str:
    client = OpenAI()
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": "You will be provided with statements, and your task is to convert them to standard English." },
            {"role": "user", "content": sentence}],
        temperature=0.7,
        max_tokens=64,
    )
    return response.choices[0].message.content

weave.init('intro-example')
correct_grammar('she no went to the market')
```
<img width="405" alt="image" src="https://github.com/wandb/weave/assets/15385696/e4d9d4d8-55d0-42cf-9602-306a280490b6">